### DISCUSSION TITLE
use pgp_sym_decrypt for columns instead of pgp_sym_encrypt for variable in the where clause of the compiled query

### DISCUSSION THREAD

**Discussion by Tuisto59** [sqlalchemy/sqlalchemy#7268] [Usage Questions] [answered]
Hi after my question in the chatroom, they say me to ask my question here.

I also since 5 days posted a question in stackoverflow, but since, nobody seems to have the answer

https://stackoverflow.com/questions/69710498/python-sqlalchemy-use-pgp-sym-decrypt-for-columns-instead-of-pgp-sym-encrypt-for

Its about TypeDecorator and the order of how SQLAlchemy compile the query when filter are added to use pgp_sym_encrypt and pgp_sym_decrypt in a SELECT + WHERE query.

Thanks a lot for your help.

yb

**Comment by zzzeek** [comment, marked as answer]
I had mentioned on gitter when I saw your message, we have a pretty comprehensive recipe for server side encryption with two different variants at https://github.com/sqlalchemy/sqlalchemy/wiki/SymmetricEncryptionServerSide , these illustrate a MySQL function but the idea would be similar with PostgreSQL functions.

looking in your SO issue though I don't know what you mean by "when we use pgp_sym_encrypt with the actual ENCRYPTION_KEY, the data will be encrypted in a new way."   if you want to filter by criteria against encrypted columns, the server side encryption feature has to return the identical encryption string every time.    otherwise you would need to manually apply the decryption function to the columns you are filtering on, something like:

    query.filter(func.my_decrypt_fn(MyModel.some_field, decryption_key, type_=String) == "some unencrypted value")

the example linked in the wiki does comparison of encrypted values by comparing the encrypted strings, im not sure why PG's function wouldn't do something similar.

**Comment by Tuisto59** [comment]
Thanks a lot, I understand now how to use the "func" and the method I create that one :

~~~
    def pgp_sym_decrypt(self, col, ):
        return func.pgp_sym_decrypt(col, self.passphrase, type_=String)
~~~

And it work !

**Comment by sluipmoord** [comment]
Hello for anyone who come across this I've implemented the a serverside encryption using pgp_sym_encrypt and pgp_sym_decrypt with a Comparator to auto decrypt the left side of the query, it uses flask_sqlalchemy plugin but should be easily modified to work without https://gist.github.com/sluipmoord/05bde6f54875283b4dfce040120ee2d6
